### PR TITLE
CDAP-6339 Add one unit test for cross namespace dataset access with a…

### PR DESCRIPTION
**Pull request by lumig242**
When cross namespace accessing was implemented, there was no authorization for datasets yet. Now add a separate unit test for authorization for cross-namespace dataset access in an application.

JIRA: https://issues.cask.co/browse/CDAP-6339
Build: http://builds.cask.co/browse/CDAP-DUT4565

### CODE REVIEW

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 1]
```diff
@@ -543,6 +549,51 @@ public Boolean call() throws Exception {
...
+    addDatasetInstance(datasetOutputSpace.toId(), "keyValueTable", "store");
+    ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.class);
+
+    // Send stream and run the flowlet as BOB. Should fail
+    StreamManager streamManager = getStreamManager(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.STREAM_NAME);
+    for (int i = 0; i < 10; i++) {
+      streamManager.send(String.valueOf(i).getBytes());
+    }
+    Authorizer authorizer = getAuthorizer();
+    authorizer.grant(AUTH_NAMESPACE, BOB, ImmutableSet.of(Action.ALL));
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:861]: why is this necessary?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 2]
```diff
@@ -543,6 +549,51 @@ public Boolean call() throws Exception {
...
+
+    DataSetManager<KeyValueTable> dataSetManager = getDataset(datasetOutputSpace.toId(), "store");
+    KeyValueTable results = dataSetManager.get();
+    for (int i = 0; i < 10; i++) {
+      byte[] key = String.valueOf(i).getBytes(Charsets.UTF_8);
+      Assert.assertArrayEquals(key, results.read(key));
+    }
+
+    flowManager.stop();
+    getNamespaceAdmin().delete(datasetOutputSpace.toId());
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:888]: also delete the app, and any other resources created

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/CrossNsDatasetAccessApp.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    setName("WriterApp");
+    setDescription("Writes from a stream to DS in another NS");
+    addStream(new Stream(STREAM_NAME));
+    addFlow(new WhoFlow());
+  }
+
+  /**
+   * Sample Flow.
+   */
+  public static final class WhoFlow extends AbstractFlow {
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/CrossNsDatasetAccessApp.java:48]: so this only tests flows, right? how about other programs?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 1 above)
Comment (lumig242) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:861]: Since we are testing dataset access. What we expect is that if Bob don't have privileges on the dataset in DATASET_SPACE, it should fail. As a result, we first grant Bob with the privileges of the original namespace with the program to make sure this won't cause the failure.

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 2 above)
Comment (lumig242) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:888]: The auth test will delete the `AuthNamespace` in the end. So all the resources this test case created in this namespace will also be deleted right?

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/CrossNsDatasetAccessApp.java (hunk 3 above)
Comment (lumig242) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/CrossNsDatasetAccessApp.java:48]: Will add for other programs.

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 4]
```diff
@@ -839,6 +843,131 @@ public Boolean call() throws Exception {
...
+    getAuthorizer().grant(AUTH_NAMESPACE, BOB, ImmutableSet.of(Action.ALL));
+    SecurityRequestContext.setUserId(BOB.getName());
+    try {
+      mrManager.start(argsForMR);
+      mrManager.waitForFinish(5, TimeUnit.MINUTES);
+      Assert.fail("Should fail here since bob does not have the privileges on input dataspace.");
+    } catch (Exception e) {
+      // Expected
+    }
+
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:922]: can we make sure that the program actually failed? e.g. that the output dataset did not get written to?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 5]
```diff
@@ -839,6 +843,131 @@ public Boolean call() throws Exception {
...
+    getAuthorizer().grant(AUTH_NAMESPACE, BOB, ImmutableSet.of(Action.ALL));
+    SecurityRequestContext.setUserId(BOB.getName());
+    try {
+      sparkManager.start();
+      sparkManager.waitForFinish(120, TimeUnit.SECONDS);
+      Assert.fail("Should fail here since bob does not have the privileges on input dataspace.");
+    } catch (Exception e) {
+      // Expected
+    }
+    // Switch back to Alice
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:966]: same comment here.

**Comment by bdmogal** [comment]
two minor comments. rest LGTM.

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 4 above)
Comment (lumig242) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:922]: Added.

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 5 above)
Comment (lumig242) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:966]: Added.
